File: sdks/java/src/main/java/io/coordin8/LeaseClient.java
package io.coordin8;

import coordin8.LeaseServiceGrpc;
import coordin8.LeaseOuterClass.GrantRequest;
import coordin8.LeaseOuterClass.RenewRequest;
import coordin8.LeaseOuterClass.CancelRequest;
import coordin8.LeaseOuterClass.Lease;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Status;
import io.grpc.StatusRuntimeException;

import java.io.Closeable;
import java.time.Instant;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

/**
 * Wraps the generated {@code LeaseService} gRPC client. There is no single
 * "the" LeaseMgr to connect to — Registry, Space, EventMgr, and TransactionMgr
 * each grant their own leases and mount {@code LeaseService} on their own
 * connection. Build one with {@link #wrap(ManagedChannel)} (given a connection
 * you already hold, e.g. {@link DjinnClient#registryLeases()}) or
 * {@link #dial(String)} (given an arbitrary grantor address, typically read
 * off a {@link LeaseRecord}'s {@code grantorHost}/{@code grantorPort}).
 */
public class LeaseClient implements Closeable {

    private final LeaseServiceGrpc.LeaseServiceBlockingStub stub;
    private final ManagedChannel ownedChannel; // null if wrapping a channel we don't own

    private LeaseClient(ManagedChannel channel, ManagedChannel ownedChannel) {
        this.stub = LeaseServiceGrpc.newBlockingStub(channel);
        this.ownedChannel = ownedChannel;
    }

    /**
     * Wrap an existing gRPC connection (e.g. one already held by a
     * {@link DjinnClient}) as a {@link LeaseClient}. Does not take ownership —
     * the caller is still responsible for closing {@code channel}; calling
     * {@link #close()} on the returned client is a no-op.
     */
    public static LeaseClient wrap(ManagedChannel channel) {
        return new LeaseClient(channel, null);
    }

    /**
     * Dial a lease grantor's address directly — typically
     * {@code grantorHost:grantorPort} read off a {@link LeaseRecord} already
     * held. The returned {@link LeaseClient} owns the connection; call
     * {@link #close()} when done.
     */
    public static LeaseClient dial(String grantorAddr) {
        int idx = grantorAddr.lastIndexOf(':');
        if (idx < 0) {
            throw new IllegalArgumentException("grantor address must be host:port, got: " + grantorAddr);
        }
        String host = grantorAddr.substring(0, idx);
        int port = Integer.parseInt(grantorAddr.substring(idx + 1));
        ManagedChannel channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
        return new LeaseClient(channel, channel);
    }

    /**
     * Release the underlying connection if this client owns one (built via
     * {@link #dial(String)}). A no-op for one built via {@link #wrap}.
     */
    @Override
    public void close() {
        if (ownedChannel != null) {
            ownedChannel.shutdown();
            try {
                ownedChannel.awaitTermination(5, TimeUnit.SECONDS);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    public LeaseRecord grant(String resourceId, long ttlSeconds) {
        Lease lease = stub.grant(GrantRequest.newBuilder()
                .setResourceId(resourceId)
                .setTtlSeconds(ttlSeconds)
                .build());
        return toRecord(lease);
    }

    public LeaseRecord renew(String leaseId, long ttlSeconds) {
        Lease lease = stub.renew(RenewRequest.newBuilder()
                .setLeaseId(leaseId)
                .setTtlSeconds(ttlSeconds)
                .build());
        return toRecord(lease);
    }

    public void cancel(String leaseId) {
        stub.cancel(CancelRequest.newBuilder()
                .setLeaseId(leaseId)
                .build());
    }

    /**
     * Start a background thread that renews the given lease at half its TTL
     * interval. Returns a {@link Closeable} — call {@code close()} to stop
     * the renewal loop (this does NOT cancel the lease itself).
     *
     * <p>{@code onFailure} is invoked on every failed renewal attempt, so the
     * caller always learns about a lost lease instead of it failing silently.
     * A transient transport error is reported but the loop keeps retrying on
     * the next tick; {@code NOT_FOUND}/{@code FAILED_PRECONDITION} (the
     * resource is genuinely gone — lease expired or never existed) is
     * reported once and stops the loop, mirroring the Go SDK's
     * {@code KeepAlive} semantics.
     *
     * @param leaseId    the lease to keep alive
     * @param ttlSeconds the TTL to request on each renewal
     * @param onFailure  callback invoked with the failure on every failed
     *                   renewal attempt (transient or terminal)
     * @return a handle to stop the keep-alive loop
     */
    public Closeable keepAlive(String leaseId, long ttlSeconds, Consumer<Throwable> onFailure) {
        ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "lease-keepalive-" + leaseId);
            t.setDaemon(true);
            return t;
        });
        long intervalSeconds = Math.max(ttlSeconds / 2, 1);
        ScheduledFuture<?> future = scheduler.scheduleAtFixedRate(() -> {
            try {
                renew(leaseId, ttlSeconds);
            } catch (Exception e) {
                onFailure.accept(e);
                if (e instanceof StatusRuntimeException sre) {
                    Status.Code code = sre.getStatus().getCode();
                    if (code == Status.Code.NOT_FOUND || code == Status.Code.FAILED_PRECONDITION) {
                        // The resource is genuinely gone (lease not found /
                        // expired) — no point retrying.
                        scheduler.shutdown();
                        return;
                    }
                }
                // Transient failure — keep trying on the next tick.
            }
        }, intervalSeconds, intervalSeconds, TimeUnit.SECONDS);

        return () -> {
            future.cancel(false);
            scheduler.shutdown();
        };
    }

    static LeaseRecord toRecord(Lease l) {
        return new LeaseRecord(
                l.getLeaseId(),
                l.getResourceId(),
                l.hasGrantedAt() ? Instant.ofEpochSecond(l.getGrantedAt().getSeconds()) : null,
                l.hasExpiresAt() ? Instant.ofEpochSecond(l.getExpiresAt().getSeconds()) : null,
                l.getTtlSeconds(),
                l.getGrantorHost(),
                l.getGrantorPort()
        );
    }

    public record LeaseRecord(
            String leaseId,
            String resourceId,
            Instant grantedAt,
            Instant expiresAt,
            long ttlSeconds,
            String grantorHost,
            int grantorPort
    ) {
        /** {@code "host:port"} for renewing/cancelling this lease — pass it to {@link LeaseClient#dial(String)}. */
        public String grantorAddr() {
            return grantorHost + ":" + grantorPort;
        }
    }
}

File: sdks/node/src/lease-client.ts
import * as grpc from "@grpc/grpc-js";
import { LeaseServiceClient } from "../gen/coordin8/lease";
import type { Lease as ProtoLease } from "../gen/coordin8/lease";
import { interceptorsFor } from "./auth";

/**
 * There is no single "the" LeaseMgr to connect to — Registry, Space, and
 * EventMgr each grant their own leases and mount LeaseService on their own
 * connection. Build a LeaseClient with `fromChannel` (given an existing
 * connection you already hold, e.g. `DjinnClient.registryLeases()`) or
 * `dial`/`dialLease` (given a grantor address, typically read off a
 * `Lease`'s `grantorHost`/`grantorPort`). See
 * .claude/plans/distributed-leasing/PRD.md.
 */
export interface Lease {
  leaseId: string;
  resourceId: string;
  grantedAt?: Date;
  expiresAt?: Date;
  ttlSeconds: number;
  /** Where to renew/cancel this lease directly — see grantorAddr(). */
  grantorHost: string;
  grantorPort: number;
}

/** Returns "host:port" for renewing/cancelling this lease — pass it to dialLease(). */
export function grantorAddr(lease: Lease): string {
  return `${lease.grantorHost}:${lease.grantorPort}`;
}

function toLease(r: ProtoLease): Lease {
  return {
    leaseId: r.leaseId,
    resourceId: r.resourceId,
    grantedAt: r.grantedAt ?? undefined,
    expiresAt: r.expiresAt ?? undefined,
    ttlSeconds: r.ttlSeconds,
    grantorHost: r.grantorHost,
    grantorPort: r.grantorPort,
  };
}

export class LeaseClient {
  private readonly stub: LeaseServiceClient;
  private readonly ownsConnection: boolean;

  private constructor(stub: LeaseServiceClient, ownsConnection: boolean) {
    this.stub = stub;
    this.ownsConnection = ownsConnection;
  }

  /**
   * Wrap an existing gRPC channel (e.g. Registry's) as a LeaseClient. Does
   * not take ownership of the channel — the caller is still responsible for
   * closing whatever owns it. This is the common case: every self-registered
   * lease is Registry-granted, so `DjinnClient.registryLeases()` builds one
   * this way.
   */
  static fromChannel(channel: grpc.Channel, interceptors: grpc.Interceptor[] = []): LeaseClient {
    const stub = new LeaseServiceClient(
      "passthrough:///djinn",
      grpc.credentials.createInsecure(),
      { channelOverride: channel, interceptors }
    );
    return new LeaseClient(stub, false);
  }

  /**
   * Dial a lease grantor directly at grantorAddr — typically
   * `grantorHost:grantorPort` read off a Lease you already hold (see
   * grantorAddr()), for leases granted by Space or EventMgr instead of
   * Registry. The returned LeaseClient owns the connection; call close()
   * when done. Pass token if the grantor has COORDIN8_JWT_SECRET set.
   */
  static dial(grantorAddr: string, token?: string): LeaseClient {
    const stub = new LeaseServiceClient(grantorAddr, grpc.credentials.createInsecure(), {
      interceptors: interceptorsFor(token),
    });
    return new LeaseClient(stub, true);
  }

  /**
   * Releases the underlying connection if this LeaseClient owns one (built
   * via dial()/dialLease()). A no-op for one built via fromChannel().
   */
  close(): void {
    if (this.ownsConnection) {
      this.stub.close();
    }
  }

  grant(resourceId: string, ttlSeconds: number): Promise<Lease> {
    return new Promise((resolve, reject) => {
      this.stub.grant({ resourceId, ttlSeconds }, (err, res) => {
        if (err || !res) return reject(err);
        resolve(toLease(res));
      });
    });
  }

  renew(leaseId: string, ttlSeconds: number): Promise<Lease> {
    return new Promise((resolve, reject) => {
      this.stub.renew({ leaseId, ttlSeconds }, (err, res) => {
        if (err || !res) return reject(err);
        resolve(toLease(res));
      });
    });
  }

  cancel(leaseId: string): Promise<void> {
    return new Promise((resolve, reject) => {
      this.stub.cancel({ leaseId }, (err) => {
        if (err) return reject(err);
        resolve();
      });
    });
  }

  /**
   * Renews leaseId in the background at half the TTL interval. Runs until
   * signal aborts or a renewal fails with the resource genuinely gone.
   *
   * Every failed renewal is reported via onFailure — a transient transport
   * error is retried on the next tick, while NOT_FOUND/FAILED_PRECONDITION
   * (the lease is genuinely gone) is reported and then stops the loop.
   * Callers that don't care about failures can omit onFailure.
   */
  keepAlive(
    leaseId: string,
    ttlSeconds: number,
    signal: AbortSignal,
    onFailure?: (err: unknown) => void
  ): void {
    const intervalMs = (ttlSeconds / 2) * 1000;
    const timer = setInterval(async () => {
      try {
        await this.renew(leaseId, ttlSeconds);
      } catch (err) {
        onFailure?.(err);
        const code = (err as grpc.ServiceError)?.code;
        if (code === grpc.status.NOT_FOUND || code === grpc.status.FAILED_PRECONDITION) {
          // The resource is genuinely gone (LeaseNotFound / LeaseExpired) —
          // no point retrying.
          clearInterval(timer);
          return;
        }
        // Transient failure — keep trying on the next tick.
      }
    }, intervalMs);
    signal.addEventListener("abort", () => clearInterval(timer));
  }
}

/**
 * Dial a lease grantor directly at grantorAddr — mirrors Go's DialLease.
 * Typically called with `grantorHost:grantorPort` read off a Lease you
 * already hold (see grantorAddr()). The returned LeaseClient owns the
 * connection; call close() when done.
 */
export function dialLease(grantorAddr: string, token?: string): LeaseClient {
  return LeaseClient.dial(grantorAddr, token);
}
